### ISSUE TITLE
Orchestration runner doesn't populate __pillar__ based on pillarenv

### ISSUE THREAD

**Issue by clallen** [saltstack/salt#42568]
I'm using a custom state module in orchestration that uses `__pillar__`:

`pools = __pillar__['autobld']['san']['pool_cus']`

When I run my orchestration, I specify the pillarenv:

> salt-run state.orch autobld.ldoms.master.storage saltenv=autobld pillarenv=autobld

and get "KeyError: 'autobld'".

However I know it is there and accessible, because if I replace the above line with this one it works:

`pools = __salt__['pillar.get']('autobld:san:pool_cus', saltenv='autobld')`

Is the `__pillar__` dict supposed to get populated based on the pillarenv you pass to state.orch, or should I just be using the "pillar.get" call and passing the env in to my custom state?

I'll be glad to provide any other information about my setup that may help.

**Comment by gtmanfred** [comment]
Can you provide the versionreport like the template for new issues requested?

There are a couple of known issues that are going to be fixed in newer versions.

Thanks,
Daniel

**Comment by clallen** [comment]
Sure, sorry about that, I knew I forgot something!

-------------------------------------------------------
Salt Version:
           Salt: 2016.11.6

Dependency Versions:
           cffi: 1.10.0
       cherrypy: unknown
       dateutil: 2.6.1
      docker-py: Not Installed
          gitdb: 2.0.2
      gitpython: 2.1.5
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: 1.5.1
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: 2.18
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.11 (default, Jul 17 2017, 21:29:29)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.3
        timelib: 0.2.4
        Tornado: 4.5.1
            ZMQ: 4.1.6

System Versions:
           dist: redhat 7.2 Maipo
        machine: x86_64
        release: 4.1.12-61.1.16.el7uek.x86_64
         system: Linux
        version: Red Hat Enterprise Linux Server 7.2 Maipo

**Comment by gtmanfred** [comment]
Can you check if 2016.11.5 works?

We had a regression in 2016.11.6, but should be fixed in 2016.11.7 by this https://github.com/saltstack/salt/pull/42211

**Comment by clallen** [comment]
I get the same behavior on 2016.11.5.  Oddly, even the `pillar.get` call doesn't work in that version (returns nothing).
Even more oddly, I can do `pillar.items` for the "autobld" env and get results, but `pillar.get` calls against that env return nothing.

**Comment by gtmanfred** [comment]
Make sure you do a `saltutil.refresh_pillar`.

pillar.items does a full pull of the pillar dictionary.  but pillar.get pulls them from the dictionary inside the current salt-minion running process, so you have to refresh the pillars to get them to update.

**Comment by clallen** [comment]
Yes, I tried that, and could never get `pillar.get` to work against anything but the base env in 11.5.  It works in 11.6.
I guess for now I'll just have to use `pillar.get` in my 11.6 code.

**Comment by gtmanfred** [comment]
Unfortunately the only thing I have been able to replicate is fixed by the PR that I linked above.

**Comment by clallen** [comment]
Ok, I'll test it in the 2016.11 branch.

**Comment by clallen** [comment]
Same issue, here's a new versions report since I'm running in a test environment different than the original one:

Salt Version:
           Salt: 2016.11.6-80-gde2f397

Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: 2.0.2
      gitpython: 2.1.5
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: 0.26.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.5 (default, Sep  5 2016, 02:30:38)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: 2.0.3
        timelib: Not Installed
        Tornado: 4.5.1
            ZMQ: 4.1.6

System Versions:
           dist: oracle 7.3
        machine: x86_64
        release: 4.1.12-94.2.1.el7uek.x86_64
         system: Linux
        version: Oracle Linux Server 7.3

Could this be something to do with using gitfs as the fileserver and pillar backend?

**Comment by gtmanfred** [comment]
Where/what do your top files look like?

Thanks,
Daniel

**Comment by clallen** [comment]
Top file locations (relative to the top of the git repo) and contents:

**files/top.sls**
```
base:
  '*':
    - grains
```

**pillar/top.sls**
```
{{saltenv}}:
  '*':
    - autobld
```

The relevant master config options:
```
fileserver_backend:
  - git

gitfs_provider: gitpython

gitfs_remotes:
  - file:///srv/salt.git:
    - root: files

ext_pillar:
  - git:
    - __env__ file:///srv/salt.git:
      - root: pillar
```

**Comment by gtmanfred** [comment]
what branch are the top files on?

**Comment by clallen** [comment]
The "autobld" branch:
```
> git branch
* autobld
  master
```

Everything I'm currently working with is on that branch.  I didn't change file/top.sls from "base" because as far as I can tell it's only used with highstate.

**Comment by gtmanfred** [comment]
I am just trying to replicate the environment so that I can test it.

Thanks,
Daniel

**Comment by clallen** [comment]
Understood, thanks!

**Comment by gtmanfred** [comment]
Ok, I have replicated this.  If I set pillarenv in /etc/salt/master, it works, but not passing pillarenv on the commandline.

```
[root@salt salt.git]# tree /srv/salt.git/
/srv/salt.git/
├── files
│   ├── deploy.sls
│   ├── _modules
│   │   └── mytest.py
│   └── top.sls
└── pillar
    ├── autobld.sls
    └── top.sls

3 directories, 5 files
[root@salt salt.git]# tail -c +0 /srv/salt.git/*/*.sls /srv/salt.git/files/_modules/*.py /etc/salt/{master,minion}
==> /srv/salt.git/files/deploy.sls <==
test:
  salt.function:
    - name: mytest.pillar
    - tgt: '*'


==> /srv/salt.git/files/top.sls <==
base:
  '*':
    - grains

==> /srv/salt.git/pillar/autobld.sls <==
foo: bar

==> /srv/salt.git/pillar/top.sls <==
{{saltenv}}:
  '*':
    - autobld

==> /srv/salt.git/files/_modules/mytest.py <==
def pillar():
    return __pillar__['foo']

==> /etc/salt/master <==
fileserver_backend:
  - git

gitfs_provider: gitpython

gitfs_remotes:
  - file:///srv/salt.git:
    - root: files

ext_pillar:
  - git:
    - __env__ file:///srv/salt.git:
      - root: pillar

==> /etc/salt/minion <==
master: localhost
[root@salt salt.git]# git branch
* autobld
  master
[root@salt salt.git]# salt-run state.orch deploy saltenv=autobld pillarenv=autobld
salt.c.gtmanfred-1263.internal_master:
----------
          ID: test
    Function: salt.function
        Name: mytest.pillar
      Result: True
     Comment: Function ran successfully. Function mytest.pillar ran on salt.c.gtmanfred-1263.internal.
     Started: 17:03:56.687855
    Duration: 89.561 ms
     Changes:
              salt.c.gtmanfred-1263.internal:
                  The minion function caused an exception: Traceback (most recent call last):
                File "/root/src/salt/salt/minion.py", line 1466, in _thread_return
                  return_data = executor.execute()
                File "/root/src/salt/salt/executors/direct_call.py", line 28, in execute
                  return self.func(*self.args, **self.kwargs)
                File "/var/cache/salt/minion/extmods/modules/mytest.py", line 2, in pillar
                  return __pillar__['foo']
                File "/root/src/salt/salt/utils/context.py", line 227, in __getitem__
                  return self._dict()[key]
              KeyError: 'foo'


Summary for salt.c.gtmanfred-1263.internal_master
------------
Succeeded: 1 (changed=1)
Failed:    0
------------
Total states run:     1
Total run time:  89.561 ms
```

@terminalmage should the pillarenv be passed through for this?

I think that this may be because `__pillar__` is populated with what the `pillarenv` is set to in the config file, since `__pillar__` is populated by the loader.  But the `salt.pillar.get` module can specify which environment to pull from.

Thanks,
Daniel

**Comment by Modulus** [comment]
Same issue with 2017.7.0
----

Salt Version:
           Salt: 2017.7.0
 
Dependency Versions:
           cffi: 1.10.0
       cherrypy: 3.2.2
       dateutil: 1.5
      docker-py: Not Installed
          gitdb: 0.5.4
      gitpython: 0.3.2 RC1
          ioflo: Not Installed
         Jinja2: 2.7.2
        libgit2: 0.25.0
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: 0.9.1
   msgpack-pure: Not Installed
 msgpack-python: 0.4.6
   mysql-python: 1.2.3
      pycparser: 2.17
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: 0.25.0
         Python: 2.7.6 (default, Oct 26 2016, 20:30:19)
   python-gnupg: Not Installed
         PyYAML: 3.10
          PyZMQ: 14.0.1
           RAET: Not Installed
          smmap: 0.8.2
        timelib: Not Installed
        Tornado: 4.2.1
            ZMQ: 4.0.5
 
System Versions:
           dist: Ubuntu 14.04 trusty
         locale: UTF-8
        machine: x86_64
        release: 3.16.0-77-generic
         system: Linux
        version: Ubuntu 14.04 trusty

**Comment by Modulus** [comment]
Also it seems like using any runner will not return any pillar data on the minion. It returns pillar data from the master.

running "salt-run pillar.show_pillar minion_id only returns the master pillar data (if you are running a minion on the master node)

If i run salt minion_id pillar.items I get the full pillar data set. This completely breaks the orchestrator when using formulas or states which are using pillar data. Please fix this

**Comment by gtmanfred** [comment]
@modulus that is a separate issue from this, and has already been solved and will be fixed in 2017.7.1

**Comment by Modulus** [comment]
@gtmanfred great, thanks :)

**Comment by clallen** [comment]
Any ideas on fixing the pillarenv arg?

**Comment by clallen** [comment]
I'm seeing the same issue with orchestrate when called from reactor, however in this case the pillar.get call does not work with pillarenv.

__files/_reactors/ldoms_provision.sls__
```
LDOMs post-install setup:
  runner.state.orchestrate:
    - mods: autobld.ldoms.provision
    - pillar:
        reactor_guestid: {{data['id']}}
    - saltenv: autobld
    - pillarenv: autobld
```

In __autobld/ldoms/provision.sls__ I tried both of these, and they use the base env rather than the pillarenv that was passed in:
```
hostmap = __salt__['pillar.get']('autobld:ldoms:hostmap')

hostmap = __pillar__['autobld']['ldoms']['hostmap']
```

**Comment by The-Loeki** [comment]
My PR #43817 fixes the behaviour; basically it defaults to `None`, passes that default on where it gets treated as an overriding param :)

**Comment by rallytime** [comment]
Closed via #43817
